### PR TITLE
Boilerplate customization

### PR DESCRIPTION
**Pull request by laerreal**
Make QDT more friendly for boilerplate customization and add an example.

Side changes:
- support v4.1.0

### v2
- fix errors
- prettify example of boilerplate customization

### CODE REVIEW

File: examples/customizing_device.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+class CustomSBDType(SysBusDeviceType):
+    """ This is an example of system bus device boilerplate extension with
+custom functionality.
+    """
+
+    def fill_source(self):
+        """ Adds a hook callback in MMIO read handler.
+        """
+
+        super().fill_source()
```
Comment (BDanAnd) [inline examples/customizing_device.py]: Error on Python2:
```
TypeError: super() takes at least 1 argument (0 given)
```

File: examples/customizing_device.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+        read_func = self.find_mmio_read_helper(0)
+
+        # We are going to pass second and following argument values of the
+        # read helper to the custom callback.
+        args2callback = read_func.args[1:]
+
+        # Insert function call statement before 2-nd statement.
+        # Note, first statement is likely a variable declaration.
+        # This is for prettiness pnly.
```
Comment (BDanAnd) [inline examples/customizing_device.py]: ```only```

File: examples/customizing_device.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        read_func.body.children.insert(1,
+            Call("custom_callback", *args2callback)
+        )
+
+    def find_mmio_read_helper(self, mmioN):
+        # Evaluate name of the read helper.
+        # This code came from `super().fill_source()`.
```
Comment (BDanAnd) [inline examples/customizing_device.py]: Fix comment after changing `super` call.

**Review by BDanAnd — COMMENTED** [review]
v1 is viewed

**Comment by laerreal** [comment]
v2 is ready

**Pull request merged**